### PR TITLE
[SYCL-MLIR] Add back the sycl commit that changes the default optimization level

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Please only review https://github.com/intel/llvm/pull/8007/commits/9899121120ddcb9a0ef70bc152e4cb12927cfaa3.

https://github.com/intel/llvm/pull/7957 merged `sycl` branch to `sycl-mlir` branch.
On of the upstream commit https://github.com/intel/llvm/commit/102e154349e169c7d5ad27c434e22f9630d300ca causes lit failures, so it was temporarily reverted in `sycl-mlir` branch.  
This PR adds back the reverted commit in `sycl-mlir` branch, and fix the failures.

Please do not squash and merge this PR.

### CODE REVIEW

File: polygeist/tools/cgeist/driver.cc [hunk 1]
```diff
@@ -946,9 +947,14 @@ splitCommandLineOptions(int argc, char **argv,
       } else if (Ref == "-fsycl-is-device") {
         SYCLIsDevice = true;
         MLIRArgs.push_back(argv[I]);
+      } else if (Ref == "--args") {
+        ClangOption = true;
+        MLIRArgs.push_back(argv[I]);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/driver.cc:952]: Where are we using this?

Comment (victor-eds) [reply polygeist/tools/cgeist/driver.cc:952]: Also, this is not mentioned in the message. What are we gonna use this for?

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/driver.cc:952]: Do you mean `--args`? Before this change, `--args` is going to the else case, which only does `MLIRArgs.push_back(argv[I]);`, so `MLIRArgs` is not changed for `--args`.

This change is part of fixing the failures caused by the upstream commit  https://github.com/intel/llvm/commit/102e154349e169c7d5ad27c434e22f9630d300ca.
`MLIRArgs` is (indirectly) used in `getOptimizationLevel()`. Before this PR, the optimization level option is not part of `MLIRArgs`, as we are only adding to `LinkageArgs`. Therefore `getOptimizationLevel()` always uses `DefaultOpt`, as no optimization option in `MLIRArgs`/`Args`.
The upstream commit modifies `DefaultOpt` in `getOptimizationLevel()`. Originally, `DefaultOpt` is changed to `-O2` in `getOptimizationLevel()` when `-fsycl-is-device` is given. Now it no longer does that, it relies on `Clang::ConstructJob` to change the default optimization to `-O2` if no optimization option is given. 
By adding the optimization level option to `MLIRArgs`, `getOptimizationLevel()` can return the correct optimization level. With incorrect optimization level, test e.g. stream-triad fails.
And we only want to do it when the optimization option is passed after `--args`, or else we get errors like
 ```
cgeist: Unknown command line argument '-O3'.  Try: '/localdisk2/waihungt/llvm/build/bin/cgeist --help'
cgeist: Did you mean '-I'?
```

Comment (victor-eds) [reply polygeist/tools/cgeist/driver.cc:952]: Nice, thanks for the explanation!

**Review by victor-eds — APPROVED** [review]

**Pull request merged**